<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <!-- For Mobile Devices -->
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
    <meta name="generator" content="Doxygen 1.8.11"/>
    <script type="text/javascript" src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
    <title>Vargas</title>
    <!--<link href="tabs.css" rel="stylesheet" type="text/css"/>-->
    <script type="text/javascript" src="dynsections.js"></script>
    <link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
    <link href="doxygen.css" rel="stylesheet" type="text/css"/>
    <link href="customdoxygen.css" rel="stylesheet" type="text/css"/>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/css/bootstrap.min.css">
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="doxy-boot.js"></script>
</head>
<body>
<nav class="navbar navbar-default" role="navigation">
    <div class="container">
        <div class="navbar-header">
            <a class="navbar-brand">Vargas 1.0 <a class="navbar-brand"
                                                                   href="https://github.com/RaviGaddipati/vargas"><font
                    size="1">Github</font></a></a>
        </div>
    </div>
</nav>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
    <div class="content" id="content">
        <div class="container">
            <div class="row">
                <div class="col-sm-12 panel panel-default" style="padding-bottom: 15px;">
                    <div style="margin-bottom: 15px;">
                        <!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Home</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>Vargas</b></li><li class="navelem"><a class="el" href="classVargas_1_1VCF.html">VCF</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="classVargas_1_1VCF-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Vargas::VCF Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;<a class="el" href="varfile_8h_source.html">varfile.h</a>&gt;</code></p>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Provides an interface to a VCF/BCF file. Core processing provided by htslib. <br />
Usage: <br />
. </p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &quot;<a class="code" href="varfile_8h.html">varfile.h</a>&quot;</span></div><div class="line"></div><div class="line"><a class="code" href="classVargas_1_1VCF.html">Vargas::VCF</a> vcf(<span class="stringliteral">&quot;variants.bcf&quot;</span>);</div><div class="line">std::vector&lt;std::string&gt; seqs = vcf.sequences(); <span class="comment">// Names of CHROM&#39;s present</span></div><div class="line"></div><div class="line">vcf.set_region(<span class="stringliteral">&quot;22:0-10,000,000&quot;</span>);</div><div class="line">vcf.create_ingroup(50); <span class="comment">// Randomly select 50% of individuals to include</span></div><div class="line"></div><div class="line">vcf.samples().size(); <span class="comment">// 4 individuals</span></div><div class="line"></div><div class="line"><span class="comment">// Allele with the maximum allele frequency for each VCF record</span></div><div class="line"><span class="keywordflow">while</span>(vcf.next()) {</div><div class="line"> <span class="keywordtype">float</span> max_af = 0;</div><div class="line"> std::string max_allele;</div><div class="line"> std::vector&lt;float&gt; &amp;freqs = vcf.frequencies();</div><div class="line"> <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; i &lt; freqs.size(); ++i) {</div><div class="line">   <span class="keywordflow">if</span> (f &gt; max_af) {</div><div class="line">    max_af = f;</div><div class="line">    max_allele = vcf.alleles[i];</div><div class="line">    }</div><div class="line">  }</div><div class="line">  std::cout &lt;&lt; <span class="stringliteral">&quot;Variant position:&quot;</span> &lt;&lt; vcf.pos() &lt;&lt; <span class="stringliteral">&quot;, Ref:&quot;</span> &lt;&lt; vcf.ref()</div><div class="line">            &lt;&lt; <span class="stringliteral">&quot;, Max AF allele:&quot;</span> &lt;&lt; max_allele &lt;&lt; <span class="stringliteral">&quot;, Freq:&quot;</span> &lt;&lt; max_af</div><div class="line">            &lt;&lt; <span class="stringliteral">&quot;, Population:&quot;</span> &lt;&lt; vcf.allele_pop(max_allele).to_string();</div><div class="line"></div><div class="line">  <span class="comment">// 4 samples/individuals = 8 genotypes</span></div><div class="line">  <span class="comment">// Variant position: 100 Ref: A, Max AF allele: G, Freq: 0.6, Population: 01100101</span></div><div class="line">}</div></div><!-- fragment --> </div><div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Inheritance diagram for Vargas::VCF:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
 <div class="center">
  <img src="classVargas_1_1VCF.png" usemap="#Vargas::VCF_map" alt=""/>
  <map id="Vargas::VCF_map" name="Vargas::VCF_map">
<area href="classVargas_1_1VariantFile.html" title="Base class for files representing variants in a reference. Provides an interface to iterate through v..." alt="Vargas::VariantFile" shape="rect" coords="0,0,121,24"/>
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classVargas_1_1VCF_1_1FormatField.html">FormatField</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the specified format field from the record.  <a href="classVargas_1_1VCF_1_1FormatField.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classVargas_1_1VCF_1_1InfoField.html">InfoField</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the specified info field from the record.  <a href="classVargas_1_1VCF_1_1InfoField.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:a8964cdea32cd38600ed42c49a25431c0"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classVargas_1_1VCF.html#a8964cdea32cd38600ed42c49a25431c0">_init</a> ()</td></tr>
<tr class="memdesc:a8964cdea32cd38600ed42c49a25431c0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Open the provided file and load the header.  <a href="#a8964cdea32cd38600ed42c49a25431c0">More...</a><br /></td></tr>
<tr class="separator:a8964cdea32cd38600ed42c49a25431c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8521a57d84e01b11616d70b99992cb81"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8521a57d84e01b11616d70b99992cb81"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classVargas_1_1VCF.html#a8521a57d84e01b11616d70b99992cb81">_load_shared</a> ()</td></tr>
<tr class="memdesc:a8521a57d84e01b11616d70b99992cb81"><td class="mdescLeft">&#160;</td><td class="mdescRight">Loads the list of alleles for the current record. <br /></td></tr>
<tr class="separator:a8521a57d84e01b11616d70b99992cb81"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa7395a947a7ab86948a42d24c6884e39"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa7395a947a7ab86948a42d24c6884e39"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classVargas_1_1VCF.html#aa7395a947a7ab86948a42d24c6884e39">_apply_ingroup_filter</a> ()</td></tr>
<tr class="memdesc:aa7395a947a7ab86948a42d24c6884e39"><td class="mdescLeft">&#160;</td><td class="mdescRight">Applies the contents of _ingroup to the header. The filter impacts all following unpacks. <br /></td></tr>
<tr class="separator:aa7395a947a7ab86948a42d24c6884e39"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="aeab1a85b52d64c9ce8f2d0c3ff4eefa7"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">Vargas::VCF::VCF </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>file</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">explicit</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">file</td><td>VCF/BCF File name </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a504daaf2589a43af1718ade82e087297"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">Vargas::VCF::VCF </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>chr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>min</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>max</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">file</td><td>VCF/BCF File name </td></tr>
    <tr><td class="paramname">chr</td><td>Chromosome </td></tr>
    <tr><td class="paramname">min</td><td>Min position, 0 indexed </td></tr>
    <tr><td class="paramname">max</td><td>Max position, 0 indexed, inclusive </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a8964cdea32cd38600ed42c49a25431c0"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int Vargas::VCF::_init </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Open the provided file and load the header. </p>
<dl class="section return"><dt>Returns</dt><dd>-1 on file open error, -2 on header read error. </dd></dl>

</div>
</div>
<a class="anchor" id="a8e2170dee0d97ff0a834fca8c89328c5"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const Population&amp; Vargas::VCF::allele_pop </td>
          <td>(</td>
          <td class="paramtype">const std::string&#160;</td>
          <td class="paramname"><em>allele</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">override</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the population set that has the allele. </p>
<p>The returned vector has the same size as number of genotypes (samples * 2). When true, that individual/phase possed that allele. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">allele</td><td>allele to get the population of </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Population of indviduals that have the allele </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#a3e0d174903339158966d1dfa202491a7">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="aea6fe878b1013e1f13f763599bf60a1a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const std::vector&lt;std::string&gt;&amp; Vargas::VCF::alleles </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>List of all the alleles in the current record. </p>
<p>The first is the reference. Allele copy number variant tags are converted, whereas other tags are substituted for the reference. </p><dl class="section return"><dt>Returns</dt><dd>vector of alleles </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#a13ce7f0f5556dbdee715dcbf4db98dac">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="a697d67df3f88da76551654e99aed5bc7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Vargas::VCF::create_ingroup </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>percent</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Create a random subset of the samples. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">percent</td><td>of samples to keep. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a5fbe4823a30390d6c0fab213e9723850"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void Vargas::VCF::create_ingroup </td>
          <td>(</td>
          <td class="paramtype">const std::vector&lt; std::string &gt; &amp;&#160;</td>
          <td class="paramname"><em>samples</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Include only the provided sample names in the <a class="el" href="classVargas_1_1Graph.html" title="Represents a Graph of the genome. ">Graph</a>. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">samples</td><td>vector of sample names </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ae025130507a4f15ada98bc6b6649ceba"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::string Vargas::VCF::file </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>File name of VCF/BCF file. </p>
<dl class="section return"><dt>Returns</dt><dd>file name </dd></dl>

</div>
</div>
<a class="anchor" id="a729df3b48fb5243870dbace1455c3f04"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename T &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;T&gt; Vargas::VCF::fmt_tag </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>tag</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get values of an arbitrary FORMAT tag. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Field format </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">tag</td><td>tag to extract </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>vector of values. </dd></dl>

</div>
</div>
<a class="anchor" id="a7a42d38c2be7c7325a8ef1294ac6c424"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const std::vector&lt; float &gt; &amp; Vargas::VCF::frequencies </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the allele frequencies of the ref and alt alleles. The ref freq is computed with 1-sum(alt_frequencies). </p>
<dl class="section return"><dt>Returns</dt><dd>const ref to vector of frequencies </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#a8543c6f95d9c3b8a26c1651a3e47d851">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="ad7c8d43c9f2f5b18c3f6c4046148369e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const std::vector&lt; std::string &gt; &amp; Vargas::VCF::genotypes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get a list of alleles for all samples (subject to sample set restriction). </p>
<p>Consecutive alleles represent phasing, e.g. all odd indexes are one phase, all even indexes are the other. Call will unpack the full record. Explicit copy number variations are replaced, other ambigous types are replaced ambiguous. A map is also built, mapping each allele to the subpopulation that has it. </p><dl class="section return"><dt>Returns</dt><dd>Vector of alleles, ordered by sample. </dd></dl>

</div>
</div>
<a class="anchor" id="a904c62f6478f6cd66bc314014bf12c8f"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool Vargas::VCF::good </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">override</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="section return"><dt>Returns</dt><dd>true if the opened file is good. </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#a18e7707d71de1c5c07ff4882d776fe5f">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="a1f8dc6f9fd93319e264b9a37dd3b3d7b"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool Vargas::VCF::good </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Check if the file is properly loaded. </p>
<dl class="section return"><dt>Returns</dt><dd>true if file is open and has a valid header. </dd></dl>

</div>
</div>
<a class="anchor" id="a445d0f1985e7e8d19b4f2b0afb1b743d"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename T &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;T&gt; Vargas::VCF::info_tag </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>tag</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get values of an arbitrary INFO tag. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Field format </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">tag</td><td>tag to retrieve values for </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>vector of values. </dd></dl>

</div>
</div>
<a class="anchor" id="ad01f1abfede454c2845965fa896b949f"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const std::vector&lt;std::string&gt;&amp; Vargas::VCF::ingroup </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="section return"><dt>Returns</dt><dd>string passed to htslib for ingroup filtering. </dd></dl>

</div>
</div>
<a class="anchor" id="aca2bfc6cb0fcf58466d3a3912a5cc73b"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::string Vargas::VCF::ingroup_str </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Ingroup parameter on BCF reading. Empty string indicates none, "-" indicates all. </p>
<dl class="section return"><dt>Returns</dt><dd>string of ingroup samples </dd></dl>

</div>
</div>
<a class="anchor" id="a13585fb58bca321d93d1125b7635f56c"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool Vargas::VCF::next </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">override</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Load the next <a class="el" href="classVargas_1_1VCF.html" title="Provides an interface to a VCF/BCF file. Core processing provided by htslib.  Usage:  ...">VCF</a> record. All information is unpacked, subject to sample set restrictions. </p>
<dl class="section return"><dt>Returns</dt><dd>false on read error or if outside restriction range. </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#aef42f1a62ec18098fd6919847fac6a46">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="a38c12c39e550e41dd427d883cd6bf2dc"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int Vargas::VCF::num_samples </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">override</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><a class="el" href="classVargas_1_1VCF.html#a38c12c39e550e41dd427d883cd6bf2dc" title="num_samples() counts each haplotype as distinct. num_samples() = samples().size() * 2 ...">num_samples()</a> counts each haplotype as distinct. <a class="el" href="classVargas_1_1VCF.html#a38c12c39e550e41dd427d883cd6bf2dc" title="num_samples() counts each haplotype as distinct. num_samples() = samples().size() * 2 ...">num_samples()</a> = <a class="el" href="classVargas_1_1VCF.html#a0830127b866d8aad310a9f1b293fd6d6" title="Get a vector of sample names. ">samples()</a>.size() * 2 </p>
<dl class="section return"><dt>Returns</dt><dd>Number of samples the <a class="el" href="classVargas_1_1VCF.html" title="Provides an interface to a VCF/BCF file. Core processing provided by htslib.  Usage:  ...">VCF</a> has. Each sample represents two genotypes. </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#a62491857f6b5aa1ba047d21ae6a4b56f">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="a58f55e3817ffa1163e189085fb403a91"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int Vargas::VCF::open </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>file</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Open the specified <a class="el" href="classVargas_1_1VCF.html" title="Provides an interface to a VCF/BCF file. Core processing provided by htslib.  Usage:  ...">VCF</a> or BCF file and load the header. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">file</td><td>filename </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>-1 on file open error, -2 on header load error, 0 otherwise </dd></dl>

</div>
</div>
<a class="anchor" id="adc6089cb802cbae83bfa07f770d75f55"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int Vargas::VCF::pos </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>0 based position, i.e. the <a class="el" href="classVargas_1_1VCF.html" title="Provides an interface to a VCF/BCF file. Core processing provided by htslib.  Usage:  ...">VCF</a> pos - 1. </p>
<dl class="section return"><dt>Returns</dt><dd>position. </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#aa1d8dee5ee6d22a453942402abac3661">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="a5cf80d8f1deafa4e96fdf175e4fe97c7"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::string Vargas::VCF::ref </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Reference allele of the current record. </p>
<dl class="section return"><dt>Returns</dt><dd>reference allele </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#af10060ef944a5bbd72cea04e56f73348">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="a0830127b866d8aad310a9f1b293fd6d6"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const std::vector&lt;std::string&gt;&amp; Vargas::VCF::samples </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">override</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get a vector of sample names. </p>
<dl class="section return"><dt>Returns</dt><dd>vector of samples </dd></dl>

<p>Implements <a class="el" href="classVargas_1_1VariantFile.html#abdbff5a077fec67cb7f255eeadc27f9d">Vargas::VariantFile</a>.</p>

</div>
</div>
<a class="anchor" id="afedd07292c13fc2c157c216675b968ef"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; std::string &gt; Vargas::VCF::sequences </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get a list of sequences in the <a class="el" href="classVargas_1_1VCF.html" title="Provides an interface to a VCF/BCF file. Core processing provided by htslib.  Usage:  ...">VCF</a> file. </p>
<dl class="section return"><dt>Returns</dt><dd>vector of sequence names </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>include/<a class="el" href="varfile_8h_source.html">varfile.h</a></li>
<li>src/<a class="el" href="varfile_8cpp.html">varfile.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- HTML footer for doxygen 1.8.8-->
<!-- start footer part -->
</div>
</div>
</div>
</div>
</div>
<hr class="footer"/>
<address class="footer">
    <center>
        Ravi Gaddipati. <a href="mailto:rgaddip1@jhu.edu"> rgaddip1@jhu.edu</a>
    </center>
    <small>
        <a href="http://www.doxygen.org/index.html">
        <img class="footer" src="doxygen.png" alt="doxygen"/>
        </a>
    </small>
</address>
</body>
</html>
