<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <!-- For Mobile Devices -->
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
    <meta name="generator" content="Doxygen 1.8.6"/>
    <script type="text/javascript" src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
    <title>Vargas</title>
    <!--<link href="tabs.css" rel="stylesheet" type="text/css"/>-->
    <script type="text/javascript" src="dynsections.js"></script>
    <link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
    <link href="doxygen.css" rel="stylesheet" type="text/css"/>
    <link href="customdoxygen.css" rel="stylesheet" type="text/css"/>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/css/bootstrap.min.css">
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="doxy-boot.js"></script>
</head>
<body>
<nav class="navbar navbar-default" role="navigation">
    <div class="container">
        <div class="navbar-header">
            <a class="navbar-brand">Vargas 1.0 <a class="navbar-brand"
                                                                   href="https://github.com/RaviGaddipati/vargas"><font
                    size="1">Github</font></a></a>
        </div>
    </div>
</nav>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
    <div class="content" id="content">
        <div class="container">
            <div class="row">
                <div class="col-sm-12 panel panel-default" style="padding-bottom: 15px;">
                    <div style="margin-bottom: 15px;">
                        <!-- end header part -->
<!-- Generated by Doxygen 1.8.6 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li class="current"><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerator</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">Vargas Documentation</div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#building">Building</a><ul><li class="level2"><a href="#dependencies">Dependencies</a></li>
<li class="level2"><a href="#compile">Compiling</a></li>
</ul>
</li>
<li class="level1"><a href="#modes">Modes of operation</a><ul><li class="level2"><a href="#define">define</a></li>
<li class="level2"><a href="#sim">sim</a></li>
<li class="level2"><a href="#align">align</a></li>
</ul>
</li>
<li class="level1"><a href="#formats">File Formats</a><ul><li class="level2"><a href="#gdef">GDEF</a></li>
<li class="level2"><a href="#reads">Reads</a></li>
<li class="level2"><a href="#aligns">Alignments</a></li>
</ul>
</li>
</ul>
</div>
<div class="textblock"><p><em>Updated: June 26, 2016</em></p>
<p>Vargas aligns short reads to a directed acyclic graph (DAG). Reads are aligned using a SIMD vectorized version of Smith-Waterman, aligning multiple reads at once. The the aligner reports the best positions of reads without traceback to keep memory usage manageable. Additionally, vargas can be used to simulate reads from a DAG while stratifying by various parameters.</p>
<h1><a class="anchor" id="building"></a>
Building</h1>
<hr/>
<p>When cloning, use the <code>--recursive</code> option to automatically retrieve dependencies. </p>
<pre class="fragment">git clone --recursive git@github.com:RaviGaddipati/vargas.git
</pre><h2><a class="anchor" id="dependencies"></a>
Dependencies</h2>
<p>Vargas relies on htslib to provide core file processing. Once cloned, the htslib is built with autoconf (version 2.63+). </p>
<pre class="fragment">cd htslib
autoconf &amp;&amp; ./configure &amp;&amp; make
</pre><p>Vargas also uses <a href="https://github.com/p12tic/libsimdpp">libsimdpp</a> for SIMD support, and <a href="https://github.com/onqtam/doctest">doctest</a>.</p>
<h2><a class="anchor" id="compile"></a>
Compiling</h2>
<p>vargas is built with cmake. </p>
<pre class="fragment">mkdir build &amp;&amp; cd build
cmake -DCMAKE_BUILD_TYPE=Release .. &amp;&amp; make
cd ..
export PATH=${PWD}/bin:$PATH
</pre><h1><a class="anchor" id="modes"></a>
Modes of operation</h1>
<hr/>
<p><code>vargas -h</code> </p>
<pre class="fragment">---------------------- vargas, Jun 24 2016. rgaddip1@jhu.edu ----------------------
Operating modes 'vargas MODE':
    define      Define a set of graphs for use with sim/align.
    sim         Simulate reads from a set of graphs.
    align       Align reads to a set of graphs.
    test        Run doctests.
    profile     Run profiles.
</pre><h2><a class="anchor" id="define"></a>
define</h2>
<p><code>vargas define -h</code> </p>
<pre class="fragment">-------------------- vargas define, Jun 24 2016. rgaddip1@jhu.edu --------------------
-f  --fasta         *&lt;string&gt; Reference filename.
-v  --var           *&lt;string&gt; VCF/BCF filename.
-t  --out           *&lt;string&gt; Output filename.
-g  --region        *&lt;string&gt; Region of graph, format CHR:MIN-MAX.
-l  --nodelen       &lt;int&gt; Max node length, default 1,000,000
-i  --ingroup       &lt;int, int...&gt; Percent ingroup subgraphs.
                    Prefix with 'c' to interpret as number of individuals.
-n  --num           &lt;int&gt; Number of unique graphs to create for all -i. Default 1.
</pre><p><code>define</code> is used to create a graph specification (*.gdf) for <code>sim</code> and <code>align</code> modes. Various subgraphs deriving off of a base graph are described. With <code>-i</code> and <code>-n</code> multiple subgraphs based on population filters are made. The original FASTA and variant files are needed in the same relative path when using the GDEF file.</p>
<p>For example: </p>
<pre class="fragment">vargas define -f hs37d5.fa -v chr22.bcf -g 22:1,000,000-2,000,000 -i 15,30 -n 2 -t test.gdf
</pre><p>Will uniquely identify (including implicit graphs) 10 graphs:</p>
<ul>
<li>2x 15% ingroup graphs</li>
<li>2x 30% ingroup graphs</li>
<li>2x 15% outgroup graphs</li>
<li>2x 30% outgroup graphs</li>
<li>Reference graph</li>
<li>Maximum allele frequency graph</li>
</ul>
<p>The ingroup determines which samples are included in the graph. If a VCF file has 50 samples, an ingroup of up to 100 individuals can be made- each haplotype is treated independently.</p>
<h2><a class="anchor" id="sim"></a>
sim</h2>
<p><code>vargas sim -h</code> </p>
<pre class="fragment">-------------------- vargas sim, Jun 24 2016. rgaddip1@jhu.edu --------------------
-g  --gdef          *&lt;string&gt; Graph definition file. Reads are simulated from Ingroups.
-t  --outfile       *&lt;string&gt; Alignment output file.
-o  --outgroup      Simulate from outgroup graphs.
-n  --numreads      &lt;int&gt; Number of reads to simulate from each subgraph, default 1000.
-m  --muterr        &lt;float, float...&gt; Read mutation error. Default 0.
-i  --indelerr      &lt;float, float...&gt; Read indel error. Default 0.
-v  --vnodes        &lt;int, int...&gt; Number of variant nodes, default any (-1).
-b  --vbases        &lt;int, int...&gt; Number of variant bases, default any (-1).
-l  --rlen          &lt;int&gt; Read length, default 50.
-a  --rate          Interpret -m, -i as rates, instead of exact number of errors.
-j  --threads       &lt;int&gt; Number of threads. 0 for maximum hardware concurrency.

Comments preceded by '#'.
-n reads are produced for each -b, -m, -i, -v combination.
</pre><p><code>sim</code> uses a GDEF file and generates <code>-n</code> reads of each combination of <code>-m</code>, <code>-i</code>, <code>-v</code>, and <code>-b</code>. <code>-m</code> Introduces mutation errors, substituting n bases with an alternate base. Likewise, <code>-i</code> will delete a base or insert a random base. With <code>-a</code>, <code>-m</code> and <code>-i</code> are interpreted as rates (0 to 1). <code>-s</code> controls which subgraphs are used to generate reads. By default, all graphs in the GDEF file are used, and their complements (outgroup).</p>
<p>For example: </p>
<pre class="fragment">vargas sim -g test.gdf -t reads -o -n 1000 -m 0,1,2 -v 0,1,2 -j 12
</pre><p>will generate 1000 reads for each combination of <code>-m</code>, <code>-v</code>, for each graph in <code>test.gdf</code> and their complements. This results in <code>3 -m * 3 -v * 1000 reads * 10 subgraphs = 90,000</code> reads.</p>
<h2><a class="anchor" id="align"></a>
align</h2>
<p><code>vargas align -h</code> </p>
<pre class="fragment">------------------- vargas align, Jun 24 2016. rgaddip1@jhu.edu -------------------
-g  --gdef          *&lt;string&gt; Graph definition file.
-r  --reads         *&lt;string, string...&gt; Read files to align.
-t  --outfile       *&lt;string&gt; Alignment output file.
-l  --rlen          &lt;int&gt; Max read length. Default 50.
-R                  Align to reference graph.
-X                  Align to maximum allele frequency graph.
-I                  Align to ingroup graph.
-O                  Align to outgroup graph.
-m  --match         &lt;int&gt; Match score, default 2.
-n  --mismatch      &lt;int&gt; Mismatch penalty, default 2.
-o  --gap_open      &lt;int&gt; Gap opening penalty, default 3.
-e  --gap_extend    &lt;int&gt; Gap extend penalty, default 1.
-j  --threads       &lt;int&gt; Number of threads. 0 for maximum hardware concurrency.

Lines beginning with '#' are ignored.
</pre><p>Reads are aligned to graphs specified in the GDEF file. <code>-R</code> Aligns the reads to the reference sequence, <code>-X</code> to the maximum allele frequency graph, <code>-I</code> to the ingroup graph, and <code>-O</code> to the outgroup graph. If a set of reads is generated from the ingroup and outgroup using <code>vargas sim -o</code>, a subsequent alignment using <code>vargas align -O</code> will align all reads to the same outgroup graph. A <code>[rxio]</code> tag at the beginning of the alignment record indicates which alignment was done.</p>
<p>For example: </p>
<pre class="fragment">vargas align  -g test.gdf -r reads -t reads.aligns -R -I -j 12
</pre><p>Will align all the reads to both the reference graph and ingroup graphs, resulting in <code>2 * number_of_reads</code> alignments.</p>
<h1><a class="anchor" id="formats"></a>
File Formats</h1>
<hr/>
<h2><a class="anchor" id="gdef"></a>
GDEF</h2>
<p>GDEF Files are produced with <code>vargas define</code>. Below is an example produced by:</p>
<p><code>vargas define -f hs37d5_22.fa -v chr22.bcf -r 22:22,000,000-23,000,000 -i c1,c5 -n 2</code> </p>
<pre class="fragment">@gdef
ref=hs37d5_22.fa;var=chr22.bcf;reg=22:22,000,000-23,000,000;nlen=1000000
i,1,0,0:00010000000000000000000000000000000000
i,1,1,0:00000000000000000000000000100000000000
i,5,0,0:00010010000000000010001000100000000000
i,5,1,0:00010000000110000000000000000001010000
</pre><p>Each gdef file begins with <code>@gdef</code>. The following line describes the source FASTA and VCF/BCF files, the region, and the maximum node length. Each following line describes an ingroup graph with the format:</p>
<p><code>i,num,id,pct:filter</code></p>
<p>where <code>num</code> indicates number of individuals if <code>pct=false</code> or percent when <code>pct=true</code>. The length of filter is equal to <code>num_samples * 2</code></p>
<h2><a class="anchor" id="reads"></a>
Reads</h2>
<p>Reads from <code>vargas sim</code> are output in FASTA format. </p>
<pre class="fragment">&gt;end=22422171 mut=0 indel=0 vnode=-1 vbase=-1 src=i,1,0,0
AAAAGGAAAAGAGATATCTGGACGCGCAGACAGAAAAAGATCCCAAAGAT
&gt;end=37431030 mut=0 indel=0 vnode=-1 vbase=-1 src=i,1,0,0
ATGTTTCGTGTCCTGGTACCAGCACAGCAACTGACATGGATGGGGGGTTA
&gt;end=44240012 mut=0 indel=0 vnode=-1 vbase=-1 src=i,1,0,0
GGGGGAGAAATCCCCAGCCTGTCCGCCACAGGAGGCCTCTGTGTTGACTC
</pre><p>Where <code>end</code> is the position of the last base in the read, <code>mut</code> is the number of mutation errors, <code>indel</code> is the number of indel erros, <code>vnode</code> is number of variant nodes traversed, and <code>vbase</code> is number of variant bases. <code>src</code> corresponds to the graph (as defined in the GDEF file, or its complement) the read came from. Parameters correspond to the profiles used to generate them, so <code>vnode=-1</code> can mean any number of variant nodes were traversed.</p>
<h2><a class="anchor" id="aligns"></a>
Alignments</h2>
<p>The output alignment format is a CSV line. </p>
<pre class="fragment">r,o,25,0,1,TTGAGCTACNGAGTGGGGCACAGCNTCCCNCAGCCCAGGCTCCACGTACG,22453471,10,0,2,2,58,22453471,1,38,22453459,32,1
r,o,25,1,1,NAGGAGCCCTATAAACNGCTTTATGTNGGCCTACTAAGCTGAGGGGCGAG,22761414,10,0,2,2,62,22761414,1,51,22761466,2,1
i,i,25,0,0,ANAGGNGCAGTGCCTGCCTTNTCCGACTCACGTACCTCNGCCCNATGTGA,22554276,10,0,2,2,70,22554276,1,36,22554404,2,1
i,i,25,1,0,TTTTGCTATTTTATGGGTCAGNGACCAATCATCATGACNATCAGATTTGA,22590921,10,0,2,2,64,22590921,1,53,22590983,1,1
</pre><p>The fields in the CSV line and there format:</p>
<ul>
<li>alignment type [rxio]</li>
<li>read origin in/outgroup [io]</li>
<li>read origin graph number [0-9]+</li>
<li>read origin graph ID [0-9]+</li>
<li>read sequence [ACGT]+</li>
<li>read origin position [0-9]+</li>
<li>read substitution errors [0-9]+</li>
<li>read indel errors [0-9]+</li>
<li>read variant nodes [0-9]+</li>
<li>read variant bases [0-9]+</li>
<li>best score [0-9]+</li>
<li>best position [0-9]+</li>
<li>best score count [0-9]+</li>
<li>second best score [0-9]+</li>
<li>second best position [0-9]+</li>
<li>second best count [0-9]+</li>
<li>read origin correlation flag [012]</li>
</ul>
<p>Where the read origin correlation flag is 1 if the read origin position had the best score, 2 if it had the second best score, otherwise 0. </p>
</div></div><!-- contents -->
<!-- HTML footer for doxygen 1.8.8-->
<!-- start footer part -->
</div>
</div>
</div>
</div>
</div>
<hr class="footer"/>
<address class="footer">
    <center>
        Ravi Gaddipati. <a href="mailto:rgaddip1@jhu.edu"> rgaddip1@jhu.edu</a>
    </center>
    <small>
        <a href="http://www.doxygen.org/index.html">
        <img class="footer" src="doxygen.png" alt="doxygen"/>
        </a>
    </small>
</address>
</body>
</html>
