<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <!-- For Mobile Devices -->
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
    <meta name="generator" content="Doxygen 1.8.11"/>
    <script type="text/javascript" src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
    <title>Vargas</title>
    <!--<link href="tabs.css" rel="stylesheet" type="text/css"/>-->
    <script type="text/javascript" src="dynsections.js"></script>
    <link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
    <link href="doxygen.css" rel="stylesheet" type="text/css"/>
    <link href="customdoxygen.css" rel="stylesheet" type="text/css"/>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/css/bootstrap.min.css">
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="doxy-boot.js"></script>
</head>
<body>
<nav class="navbar navbar-default" role="navigation">
    <div class="container">
        <div class="navbar-header">
            <a class="navbar-brand">Vargas 1.0 <a class="navbar-brand"
                                                                   href="https://github.com/RaviGaddipati/vargas"><font
                    size="1">Github</font></a></a>
        </div>
    </div>
</nav>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
    <div class="content" id="content">
        <div class="container">
            <div class="row">
                <div class="col-sm-12 panel panel-default" style="padding-bottom: 15px;">
                    <div style="margin-bottom: 15px;">
                        <!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li class="current"><a href="index.html"><span>Home</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">Vargas Documentation</div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#building">Building</a><ul><li class="level2"><a href="#dependencies">Dependencies</a></li>
<li class="level2"><a href="#compile">Compiling</a></li>
</ul>
</li>
<li class="level1"><a href="#modes">Modes of operation</a><ul><li class="level2"><a href="#define">define</a></li>
<li class="level2"><a href="#sim">sim</a></li>
<li class="level2"><a href="#align">align</a></li>
<li class="level2"><a href="#convert">convert</a></li>
</ul>
</li>
</ul>
</div>
<div class="textblock"><p><em>Updated: June 26, 2016</em></p>
<p>Vargas aligns short reads to a directed acyclic graph (DAG). Reads are aligned using a SIMD vectorized version of Smith-Waterman, aligning multiple reads at once. The the aligner reports the best positions of reads without traceback to keep memory usage manageable. Additionally, vargas can be used to simulate reads from a DAG while stratifying by various parameters.</p>
<h1><a class="anchor" id="building"></a>
Building</h1>
<hr/>
<p>When cloning, use the <code>--recursive</code> option to automatically retrieve dependencies. </p><pre class="fragment">git clone --recursive git@github.com:RaviGaddipati/vargas.git
</pre><h2><a class="anchor" id="dependencies"></a>
Dependencies</h2>
<p>Vargas relies on htslib to provide core file processing. Once cloned, the htslib is built with autoconf (version 2.63+). </p><pre class="fragment">cd htslib
autoconf &amp;&amp; ./configure &amp;&amp; make
</pre><p>Vargas also uses <a href="https://github.com/p12tic/libsimdpp">libsimdpp</a> for SIMD support, and <a href="https://github.com/onqtam/doctest">doctest</a>.</p>
<h2><a class="anchor" id="compile"></a>
Compiling</h2>
<p>vargas is built with cmake. </p><pre class="fragment">mkdir build &amp;&amp; cd build
cmake -DCMAKE_BUILD_TYPE=Release .. &amp;&amp; make
cd ..
export PATH=${PWD}/bin:$PATH
</pre><h1><a class="anchor" id="modes"></a>
Modes of operation</h1>
<hr/>
<p><code>vargas -h</code></p>
<div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;---------------------- Vargas, Aug 31 2016. rgaddip1@jhu.edu ----------------------</div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;Operating modes &#39;Vargas MODE&#39;:</div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;    define      Define a set of graphs for use with sim/align.</div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;    sim         Simulate reads from a set of graphs.</div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;    align       Align reads to a set of graphs.</div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;    split       Split a SAM file into multiple files.</div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;    merge       Merge SAM files.</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;    convert     Convert a SAM file to a CSV file.</div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;    test        Run doctests.</div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;    profile     Run profiles.</div></div><!-- fragment --><h2><a class="anchor" id="define"></a>
define</h2>
<p><code>vargas define -h</code></p>
<div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;-------------------- Vargas define, Aug 31 2016. rgaddip1@jhu.edu --------------------</div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;-f  --fasta         *&lt;string&gt; Reference filename.</div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;-v  --vcf           &lt;string&gt; VCF or BCF file.</div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;-k  --ksnp          &lt;string&gt; KSNP File. Only one of -v or -k should be defined.</div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;-g  --region        *&lt;string&gt; Region of graph, format CHR:MIN-MAX.</div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;-l  --nodelen       &lt;int&gt; Max node length, default 1,000,000</div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;-s  --subgraph      *&lt;string&gt; Subgraph definition file. Default stdin.</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;-t  --out           &lt;string&gt; Output file, default stdout.</div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;-d  --dot           &lt;string&gt; Output DOT graph to file.</div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;</div><div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;Subgraphs are defined using the format &quot;label=N[%]&quot;.</div><div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;&#39;N&#39; is the number of samples / percentage of samples selected.</div><div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;The samples are selected from the parent graph, scoped with &#39;:&#39;.</div><div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;The base graph is implied as the root for all labels. Example:</div><div class="line"><a name="l00015"></a><span class="lineno">   15</span>&#160;a=50;a:b=10;~a:c=5</div><div class="line"><a name="l00016"></a><span class="lineno">   16</span>&#160;&#39;~&#39; indicates the complement graph. &#39;BASE&#39; refers the the whole graph.</div></div><!-- fragment --><p><code>define</code> is used to create a graph specification (*.gdf) for <code>sim</code> and <code>align</code> modes. Various subgraphs deriving off of a base graph are described. With <code>-i</code> and <code>-n</code> multiple subgraphs based on population filters are made. The original FASTA and variant files are needed in the same relative path when using the GDEF file.</p>
<p>For example: </p><pre class="fragment">vargas define -f hs37d5.fa -v chr22.bcf -g 22:1,000,000-2,000,000 -i 15,30 -n 2 -t test.gdf
</pre><p>Will uniquely identify (including implicit graphs) 10 graphs:</p>
<ul>
<li>2x 15% ingroup graphs</li>
<li>2x 30% ingroup graphs</li>
<li>2x 15% outgroup graphs</li>
<li>2x 30% outgroup graphs</li>
<li>Reference graph</li>
<li>Maximum allele frequency graph</li>
</ul>
<p>The ingroup determines which samples are included in the graph. If a VCF file has 50 samples, an ingroup of up to 100 individuals can be made- each haplotype is treated independently.</p>
<h2><a class="anchor" id="sim"></a>
sim</h2>
<p><code>vargas sim -h</code></p>
<div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;-------------------- Vargas sim, Aug 31 2016. rgaddip1@jhu.edu --------------------</div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;-g  --gdef          *&lt;string&gt; Graph definition file. Default stdin.</div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;-s  --sub           *&lt;string;string; ...&gt; list of graphs to simulate from.</div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;-f  --file          -s specifies a file name.</div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;-t  --out           &lt;string&gt; Output file. Default stdout.</div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;-n  --numreads      &lt;int&gt; Number of reads to simulate from each profile, default 1000.</div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;-m  --muterr        &lt;int/float, int/float...&gt; Read mutation error. Default 0.</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;-i  --indelerr      &lt;int/float, int/float...&gt; Read indel error. Default 0.</div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;-d  --vnodes        &lt;int, int...&gt; Number of variant nodes, default any (*).</div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;-b  --vbases        &lt;int, int...&gt; Number of variant bases, default any (*).</div><div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;-l  --rlen          &lt;int&gt; Read length, default 50.</div><div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;-a  --rate          Interpret -m, -i as rates, instead of exact number of errors.</div><div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;-j  --threads       &lt;int&gt; Number of threads. 0 for maximum hardware concurrency.</div><div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;</div><div class="line"><a name="l00015"></a><span class="lineno">   15</span>&#160;-n reads are produced for each -m, -i, -v, -b combination. If set to &#39;*&#39;, any value is accepted.</div></div><!-- fragment --><p><code>sim</code> uses a GDEF file and generates <code>-n</code> reads of each combination of <code>-m</code>, <code>-i</code>, <code>-v</code>, and <code>-b</code>. <code>-m</code> Introduces mutation errors, substituting n bases with an alternate base. Likewise, <code>-i</code> will delete a base or insert a random base. With <code>-a</code>, <code>-m</code> and <code>-i</code> are interpreted as rates (0 to 1). <code>-s</code> controls which subgraphs are used to generate reads. By default, all graphs in the GDEF file are used, and their complements (outgroup).</p>
<p>For example: </p><pre class="fragment">vargas sim -g test.gdf -t reads -o -n 1000 -m 0,1,2 -v 0,1,2 -j 12
</pre><p>will generate 1000 reads for each combination of <code>-m</code>, <code>-v</code>, for each graph in <code>test.gdf</code> and their complements. This results in <code>3 -m * 3 -v * 1000 reads * 10 subgraphs = 90,000</code> reads.</p>
<h2><a class="anchor" id="align"></a>
align</h2>
<p><code>vargas align -h</code></p>
<div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;------------------- Vargas align, Aug 31 2016. rgaddip1@jhu.edu -------------------</div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;-g  --gdef          *&lt;string&gt; Graph definition file.</div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;-r  --reads         *&lt;string, string...&gt; Read files to align. Default stdin.</div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;-a  --align         *&lt;string&gt; Alignment targets.</div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;-f  --file          -a specifies a file name.</div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;-t  --out           *&lt;string&gt; Alignment output file, default stdout.</div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;-l  --rlen          &lt;int&gt; Max read length. Default 50.</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;-m  --match         &lt;int&gt; Match score, default 2.</div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;-n  --mismatch      &lt;int&gt; Mismatch penalty, default 2.</div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;-o  --gap_open      &lt;int&gt; Gap opening penalty, default 3.</div><div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;-e  --gap_extend    &lt;int&gt; Gap extend penalty, default 1.</div><div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;-j  --threads       &lt;int&gt; Number of threads. 0 for maximum hardware concurrency.</div><div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;                            Optimal reads per subgraph: n * j * 16</div></div><!-- fragment --><p>Reads are aligned to graphs specified in the GDEF file. <code>-R</code> Aligns the reads to the reference sequence, <code>-X</code> to the maximum allele frequency graph, <code>-I</code> to the ingroup graph, and <code>-O</code> to the outgroup graph. If a set of reads is generated from the ingroup and outgroup using <code>vargas sim -o</code>, a subsequent alignment using <code>vargas align -O</code> will align all reads to the same outgroup graph. A <code>[rxio]</code> tag at the beginning of the alignment record indicates which alignment was done.</p>
<p>For example: </p><pre class="fragment">vargas align  -g test.gdf -r reads -t reads.aligns -R -I -j 12
</pre><p>Will align all the reads to both the reference graph and ingroup graphs, resulting in <code>2 * number_of_reads</code> alignments.</p>
<h2><a class="anchor" id="convert"></a>
convert</h2>
<p><code>vargas convert -h</code></p>
<div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;-------------------- Vargas convert, Aug 31 2016. rgaddip1@jhu.edu --------------------</div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;-s  --sam          &lt;string&gt; SAM input file. Default stdin.</div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;-f  --format       *&lt;string,string...&gt; Specify tags per column. Case sensitive.</div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;</div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;Output printed to stdout.</div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;Requred column names: QNAME, FLAG, RNAME, POS, MAPQ, CIGAR, RNEXT, PNEXT, TLEN, SEQ, QUAL</div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;Prefix with &quot;RG:&quot; to obtain a value from the associated read group.</div></div><!-- fragment --><p>Convert a SAM file into a CSV file, outputting the specified fields. </p>
</div></div><!-- contents -->
<!-- HTML footer for doxygen 1.8.8-->
<!-- start footer part -->
</div>
</div>
</div>
</div>
</div>
<hr class="footer"/>
<address class="footer">
    <center>
        Ravi Gaddipati. <a href="mailto:rgaddip1@jhu.edu"> rgaddip1@jhu.edu</a>
    </center>
    <small>
        <a href="http://www.doxygen.org/index.html">
        <img class="footer" src="doxygen.png" alt="doxygen"/>
        </a>
    </small>
</address>
</body>
</html>
